### ISSUE TITLE
io.fabric8.kubernetes.client.Config should expose all and the current context defined in kubeconfig

### ISSUE THREAD

**Issue by adietish** [fabric8io/kubernetes-client#2215]
Ex. If you need to list all clusters that are present in your kubeconfig there's no other way than to (re-) read and (re-) parse ```~/.kube/config```. ```io.fabric8.kubernetes.client.Config``` should therefore expose this info that is present in ```io.fabric8.kubernetes.api.model.Config```.
There are also more values in the kubeconfig that a 3rd party programmer might want/need (ex. current cluster, etc.).

**Comment by adietish** [comment]
I can contribute a PR, all I need is to know if you're ok with this tight change or if you'd alternatively prefer to expose ```io.fabric8.kubernetes.api.model.Config``` which would allow 3rd party programmers to expose whatever they want.

**Comment by rohanKanojia** [comment]
@adietish : @manusa @oscerd thoughts?? Hiding `io.fabric8.kubernetes.client.Config` and exposing specific values sound good to me. Maybe it would be less confusing.

**Comment by oscerd** [comment]
+1

**Comment by adietish** [comment]
@rohanKanojia you mean hiding ```io.fabric8.kubernetes.api.model.Config``` (the resource class which is deserialized) which is currently hidden behind the facade ```io.fabric8.kubernetes.client.Config```? 
If this is what you're after, works perfectly for me.

**Comment by manusa** [comment]
> Hiding io.fabric8.kubernetes.client.Config and exposing specific values sound good to me. Maybe it would be less confusing.

I have the same confusion as @adietish. Do you mean exposing certain fields (supposedly parsed from [io.fabric8...model.Config](https://github.com/fabric8io/kubernetes-client/blob/160446a87ecce3812d3f6f90281b59f95bb5847b/kubernetes-model/kubernetes-model-core/src/main/resources/schema/kube-schema.json#L8140)) in [io.fabric8...client.Config](https://github.com/fabric8io/kubernetes-client/blob/160446a87ecce3812d3f6f90281b59f95bb5847b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java#L58)? Yes, I think it's safe enough (and probably quite useful) to expose the current configuration values.


Back to the original issue, I'm not sure that what you can find in `client.Config` matches what you find in your deserialized local `~/.kube/config` into `model.Config. In case of a completely autoconfigured `KuberenetesClient`, this should of course be the same. But in case the client was configured manually, the configuration settings won't match with those in your local `~/.kube/config` (maybe the file was never parsed > `kubernetes.auth.tryKubeConfig=false`).

I see there is helper a method in [KubeConfigUtils](https://github.com/fabric8io/kubernetes-client/blob/160446a87ecce3812d3f6f90281b59f95bb5847b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/KubeConfigUtils.java#L38) to deserialize configurations. Even after exposing the active configuration values/settings in `client.Config`, I'm not sure, but maybe it's safer to deserialize again the file to be sure you're retrieving the right information.

**Comment by adietish** [comment]
> > Hiding io.fabric8.kubernetes.client.Config and exposing specific values sound good to me. Maybe it would be less confusing.
> 
> Yes, I think it's safe enough (and probably quite useful) to expose the current configuration values.

ok, great.

> But in case the client was configured manually, the configuration settings won't match with those in your local `~/.kube/config` (maybe the file was never parsed > `kubernetes.auth.tryKubeConfig=false`).

Thx for the insight, wasn't aware, that's a very helpful pointer. What are the consequences exactly?

> 
> I see there is helper a method in [KubeConfigUtils](https://github.com/fabric8io/kubernetes-client/blob/160446a87ecce3812d3f6f90281b59f95bb5847b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/internal/KubeConfigUtils.java#L38) to deserialize configurations. Even after exposing the active configuration values/settings in `client.Config`, I'm not sure, but maybe it's safer to deserialize again the file to be sure you're retrieving the right information.

I'd think that one would be interested in the overriding values, not the overridden ones. And for the ones that might not be available (```kubernetes.auth.tryKubeConfig=false```) load&parse as you pointed out?

**Comment by manusa** [comment]
Sorry, I think I misunderstood the question/issue.

I'm understanding 2 different things here:

1) "client.Config should expose all clusters defined in `~/.kube/config`" For this case what I'm saying is that maybe in the first place that config was never parsed to start with. So if you want to read a value that's there, the safest option is to parse the file again.

2) "client.Config should expose all other values": I would rephrase that to should expose all values active in the current client configuration. e.g. I want to know which is the current masterUrl in use.

I don't know if this is what you're actually asking/proposing.

Anyway, `client.Config` should definitely expose current configuration values (if it doesn't already).

**Comment by adietish** [comment]
> 1. "client.Config should expose all clusters defined in `~/.kube/config`" For this case what I'm saying is that maybe in the first place that config was never parsed to start with. So if you want to read a value that's there, the safest option is to parse the file again.

Would you still agree in saying that the kubernetes-client could do this if requested so that user programmers would not have to replicate the code?

**Comment by adietish** [comment]
Hi @manusa 

> 2. "client.Config should expose all other values": I would rephrase that to should expose all values active in the current client configuration. e.g. I want to know which is the current masterUrl in use.

"all active values" is too wide in the context of my requirements. I currently need to list all contexts that exist in the kubeconfig and also to know which one of these is the the current one. ```KubeConfigUtils.getCurrentContext(config)``` currently allows to extract the current ```io.fabric8.kubernetes.api.model.Context```. Context unfortunately isn't good enough for me, I need to display the name of it and thus need ``` io.fabric8.kubernetes.api.model.NamedContext``` instead.

**Comment by manusa** [comment]
Hi @adietish 
Sorry for the delay, just saw your updates.

> Would you agree that the kubernetes-client should do this (if requested) so that user programmers would not have to replicate the code?

Yes, sure. Whatever improves developer's experience is welcome ;) Especially if it reuses internal methods, so having a clean public API to expose that seems pretty good (IMHO).

In relation to https://github.com/fabric8io/kubernetes-client/issues/2215#issuecomment-643283560
I think changing the method return type (as you did in #2290) is the right approach to fulfill your requirements.

**Comment by adietish** [comment]
@manusa here's a naive implementation: #2290. Please review
